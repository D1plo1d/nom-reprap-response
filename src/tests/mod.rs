use insta::assert_debug_snapshot;
use std::collections::BTreeMap;
use colored::Colorize;

use eyre::{
    eyre,
    // Context as _,
};

pub use super::{
    parse_response,
    Response,
};

fn responses_for(v: &str) -> eyre::Result<Vec<Response>> {
    std::iter::repeat(true)
        .scan(v, |acc, _| {
            let mut try_parse = move || {
                if acc.len() == 0 {
                    return Ok(None)
                }

                let (
                    remaining,
                    (parsed, response)
                ) = parse_response(acc)
                    .map_err(|err| {
                        eyre!(
                            "[{}]\n Error: {:?}. Input: {:?}\n",
                            "ERROR".red(),
                            err,
                            acc,
                        )
                    })?;

                if let Response::Unknown = response {
                    Err(eyre!(
                        "[{}]\nParsed: {:?} . Input: {:?}\n",
                        "UNKNOWN RESPONSE".red(),
                        parsed,
                        acc,
                    ))?;
                };

                *acc = remaining;

                Ok(Some(response))
            };

            try_parse().transpose()
        })
        .collect::<eyre::Result<Vec<Response>>>()
}

fn snapshot_test_responses(data: &str) -> eyre::Result<()> {
    let data: BTreeMap<String, BTreeMap<String, String>> = toml::from_str(data)?;

    type SectionResponseTree = BTreeMap<String, Vec<Response>>;
    type Snapshot = BTreeMap<String, SectionResponseTree>;

    let responses: eyre::Result<Snapshot> = data
        .iter()
        .map(|(section_title, scenarios)| {
            let section_title = section_title.clone();

            let section_responses = scenarios.clone()
                .iter()
                .map(|(k, v)| {
                    let title = format!("{}::{}", &section_title, k);

                    print!("\nTesting {:?} ", title);
                    let res = responses_for(&v)?;

                    println!("[{}]", "DONE".green());
                    println!("Got: {:?}\n", res);

                    Ok((k.to_owned(), res))
                })
                .collect::<eyre::Result<SectionResponseTree>>()?;

            Ok((section_title, section_responses))
        })
        .collect();

    assert_debug_snapshot!(responses?);

    Ok(())
}

#[test]
fn ender3_marlin_2019_firmware() -> eyre::Result<()> {
    let data = include_str!("data/ender3_marlin_2019_firmware.toml");
    snapshot_test_responses(data)
}

#[test]
fn ultimaker2_marlin_dbg_2019_firmware() -> eyre::Result<()> {
    let data = include_str!("data/ultimaker2_marlin_dbg_2019_firmware.toml");
    snapshot_test_responses(data)
}
