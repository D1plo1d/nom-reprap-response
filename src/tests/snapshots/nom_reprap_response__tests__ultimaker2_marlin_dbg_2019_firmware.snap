---
source: src/tests/mod.rs
expression: responses?
---
{
    "errors": {
        "unknown_command": [
            Echo(
                "Unknown command: \"N3 m23 file.txt*65\"",
            ),
            Ok(
                None,
            ),
        ],
    },
    "firmware": {
        "m115_firmware_info": [
            Echo(
                "FIRMWARE_NAME:Marlin Ultimaker2; Sprinter/grbl mashup for gen6 FIRMWARE_URL:http://github.com/Ultimaker PROTOCOL_VERSION:1.0 MACHINE_TYPE:Ultimaker EXTRUDER_COUNT:1",
            ),
            Ok(
                None,
            ),
        ],
    },
    "greeting": {
        "greeting_with_sd_card": [
            Greeting,
            Echo(
                "Marlin 1.0.0",
            ),
            Echo(
                "Last Updated: Nov 17 2017 23:41:38 | Author: Version DEV",
            ),
            Debug(
                "Compiled: Nov 17 2017",
            ),
            Echo(
                "Free Memory: 2052  PlannerBufferBytes: 1232",
            ),
            Echo(
                "Stored settings retrieved",
            ),
            Echo(
                "Steps per unit:",
            ),
            Echo(
                "M92 X80.00 Y80.00 Z200.00 E311.00",
            ),
            Echo(
                "Maximum feedrates (mm/s):",
            ),
            Echo(
                "M203 X300.00 Y300.00 Z40.00 E45.00",
            ),
            Echo(
                "Maximum Acceleration (mm/s2):",
            ),
            Echo(
                "M201 X9000 Y9000 Z100 E10000",
            ),
            Echo(
                "Acceleration: S=acceleration, T=retract acceleration",
            ),
            Echo(
                "M204 S3000.00 T3000.00",
            ),
            Echo(
                "Advanced variables: S=Min feedrate (mm/s), T=Min travel feedrate (mm/s), B=minimum segment time (ms), X=maximum XY jerk (mm/s),  Z=maximum Z jerk (mm/s),  E=maximum E jerk (mm/s)",
            ),
            Echo(
                "M205 S0.00 T0.00 B20000 X20.00 Z0.40 E5.00",
            ),
            Echo(
                "Home offset (mm):",
            ),
            Echo(
                "M206 X0.00 Y0.00 Z-11.45",
            ),
            Echo(
                "PID settings:",
            ),
            Echo(
                "M301 P10.03 I1.50 D70.00",
            ),
        ],
        "greeting_without_sd_card": [
            Greeting,
            Echo(
                "Marlin 1.0.0",
            ),
            Echo(
                "Last Updated: Nov 17 2017 23:41:38 | Author: Version DEV",
            ),
            Debug(
                "Compiled: Nov 17 2017",
            ),
            Echo(
                "Free Memory: 2052  PlannerBufferBytes: 1232",
            ),
            Echo(
                "Stored settings retrieved",
            ),
            Echo(
                "Steps per unit:",
            ),
            Echo(
                "M92 X80.00 Y80.00 Z200.00 E311.00",
            ),
            Echo(
                "Maximum feedrates (mm/s):",
            ),
            Echo(
                "M203 X300.00 Y300.00 Z40.00 E45.00",
            ),
            Echo(
                "Maximum Acceleration (mm/s2):",
            ),
            Echo(
                "M201 X9000 Y9000 Z100 E10000",
            ),
            Echo(
                "Acceleration: S=acceleration, T=retract acceleration",
            ),
            Echo(
                "M204 S3000.00 T3000.00",
            ),
            Echo(
                "Advanced variables: S=Min feedrate (mm/s), T=Min travel feedrate (mm/s), B=minimum segment time (ms), X=maximum XY jerk (mm/s),  Z=maximum Z jerk (mm/s),  E=maximum E jerk (mm/s)",
            ),
            Echo(
                "M205 S0.00 T0.00 B20000 X20.00 Z0.40 E5.00",
            ),
            Echo(
                "Home offset (mm):",
            ),
            Echo(
                "M206 X0.00 Y0.00 Z-11.45",
            ),
            Echo(
                "PID settings:",
            ),
            Echo(
                "M301 P10.03 I1.50 D70.00",
            ),
            Echo(
                "SD init fail",
            ),
        ],
    },
    "m21_init_sd_card": {
        "m21_no_sd_card": [
            Echo(
                "SD init fail",
            ),
            Ok(
                None,
            ),
        ],
    },
    "m23_select_sd_file": {
        "m23_no_sd_card": [
            Ok(
                None,
            ),
        ],
    },
    "m28_start_sd_write": {
        "m28_no_sd_card": [
            Ok(
                None,
            ),
        ],
    },
    "m30_delete_sd_file": {
        "m30_no_file": [
            Ok(
                None,
            ),
        ],
    },
    "movement_gcodes": {
        "g1": [
            Ok(
                None,
            ),
        ],
        "g28": [
            Ok(
                None,
            ),
        ],
        "g90": [
            Ok(
                None,
            ),
        ],
        "g91": [
            Ok(
                None,
            ),
        ],
    },
    "polling_mcodes": {
        "m105": [
            Ok(
                Some(
                    ActualTemperatures(
                        [
                            (
                                "e0",
                                38.2,
                            ),
                            (
                                "b",
                                26.6,
                            ),
                        ],
                    ),
                ),
            ),
        ],
        "m114": [
            Feedback(
                Positions(
                    Positions {
                        target_positions: Some(
                            [
                                (
                                    "x",
                                    0.0,
                                ),
                                (
                                    "y",
                                    0.0,
                                ),
                                (
                                    "z",
                                    0.0,
                                ),
                                (
                                    "e0",
                                    0.0,
                                ),
                            ],
                        ),
                        actual_positions: [
                            (
                                "x",
                                0.0,
                            ),
                            (
                                "y",
                                0.0,
                            ),
                            (
                                "z",
                                0.0,
                            ),
                        ],
                    },
                ),
            ),
        ],
        "m155_temperature_auto_report": [
            Ok(
                None,
            ),
        ],
    },
    "temperature_control": {
        "m109": [
            Feedback(
                ActualTemperatures(
                    [
                        (
                            "e0",
                            24.7,
                        ),
                    ],
                ),
            ),
            Feedback(
                ActualTemperatures(
                    [
                        (
                            "e0",
                            24.4,
                        ),
                    ],
                ),
            ),
            Feedback(
                ActualTemperatures(
                    [
                        (
                            "e0",
                            24.4,
                        ),
                    ],
                ),
            ),
        ],
    },
}
